### DISCUSSION TITLE
Translatable popup Relation managers

### DISCUSSION THREAD

**Discussion by tombien** [filamentphp/filament#1514] [Showcase]
Hello,
I am working on solution, how to edit translatable fields in popup relation managers.
I made a trait, that emit event, when you switch language on resource page. This trait is used in Edit page (EditRecord).
And I made a second trait, that is used in RealationManager. This trait listen to "parents" event language change and prepare correct translated data before form fill. Handles updating and creation of record.
I will share this solution. I need to solve some stuff for now (show right data in Relation Manager table on Edit page).

**Comment by tombien** [comment]
Hello again,
for now I have this:

Resource Edit page emits locale change and relation manager listen to this, prepare data and handle update and creation of translatable related resource.

So from user view - you have an edit page of resource with table with related records. You can switch language from EN to CS for example and related record opens popup editor in this current translation.

app/Filament/Traits/TranslatableWithEmit.php - you should use this in edit pages (Filament/Resources/Pages/Edit.....php) instead of original Translatable trait
```
namespace App\Filament\Traits;

trait TranslatableWithEmit
{
    use \Filament\Resources\Pages\EditRecord\Concerns\Translatable{
        updatedActiveFormLocale as parentUpdatedActiveFormLocale;
        setActiveFormLocale as parentSetActiveFormLocale;
    }

    public function __construct() {
        //keep active locale if browser refresh
        $this->queryString[] = 'activeFormLocale';
        
        parent::__construct();
    }

    public function setActiveFormLocale() {
        $this->parentSetActiveFormLocale();
    }

    protected function updatedActiveFormLocale($value): void
    {
        $this->emit('activeFormLocaleUpdated', $value);
        
        $this->parentUpdatedActiveFormLocale();
    }
}
```
and 

app/Filament/Traits/HasTranslatableOwner.php
use this trait in RelationManagers with translatable model which belongs to model with TranslatableWithEmit resource.

```
namespace App\Filament\Traits;

use Illuminate\Database\Eloquent\Model;

trait HasTranslatableOwner
{
    public $activeFormLocale = null;

    public function __construct() {
        $this->listeners['activeFormLocaleUpdated'] = 'setActiveFormLocale';
    }

    public function setActiveFormLocale($value = null): void {
        if(is_null($value)) $value = $this->ownerRecord->getLocale();
        $this->activeFormLocale = $value;
    }

    protected function mutateFormDataBeforeFill(array $data): array
    {
        if ($this->activeFormLocale === null) {
            $this->setActiveFormLocale();
        }

        foreach ($this->getMountedTableActionRecord()->getTranslatableAttributes() as $attribute) {
            $data[$attribute] = $this->getMountedTableActionRecord()->getTranslation($attribute, $this->activeFormLocale);
        }

        return $data;
    }

    protected function handleRecordUpdate(Model $record, array $data): Model
    {
        $record->setLocale($this->activeFormLocale)->fill($data)->save();

        return $record;
    }

    protected function handleRecordCreation(array $data): Model
    {
        if ($this->activeFormLocale === null) {
            $this->setActiveFormLocale();
        }

        $record = $this->getRelatedModel()::getModel()::usingLocale(
            $this->activeFormLocale
        )->fill($data);

        $this->ownerRecord->{$this->getRelationshipName()}()->save($record);

        return $record;
    }
}
```

**Reply by tomas-doudera** [reply]
Thank you it helps me.

**Reply by ArtMin96** [reply]
Thanks @tombien . Do you have an updated code?
![ogvacation](https://user-images.githubusercontent.com/29732308/177733671-cf6cb314-a29a-4212-a884-7e080ee38578.png)

**Reply by tombien** [reply]
Do you have correct setup on your model? You need to setup translatable fields.

**Reply by ArtMin96** [reply]
@tombien I think yes )

Country Model

![VxnjnQ6](https://user-images.githubusercontent.com/29732308/177753947-f391b651-a456-43d9-bbaf-3462c2af1057.png)

State Model

![i6BEgLr](https://user-images.githubusercontent.com/29732308/177755135-a596af54-2de9-4990-a8de-34ae3c8fcb20.png)


Country EditResource

![Xmj81uo](https://user-images.githubusercontent.com/29732308/177754019-58f270b8-fd18-4b85-85fa-e6e3b8872c1e.png)

StateRelationManager

![jyeozim](https://user-images.githubusercontent.com/29732308/177754108-02f39303-3139-4a9a-a218-de06dca3539c.png)

**Comment by lordjoo** [comment]
Even after all of this still the listeners doesn't fire

**Comment by thijskuilman** [comment]
The suggest work-around doesn't seem to work for me. Has there been any update regarding this issue in the meantime?

**Comment by carlituxman** [comment]
I just do it using Customizing data before filling the form:
https://filamentphp.com/docs/2.x/admin/resources/relation-managers#customizing-data-before-filling-the-form
and Customizing data before saving:
https://filamentphp.com/docs/2.x/admin/resources/relation-managers#customizing-data-before-saving-1

First, create a Component Tab with every tab for translation at form:

``` 
Forms\Components\Tabs::make(__('Translations'))
                    ->tabs(function (): array {
                        $tabs = [];
                        foreach( $mylocales as $locale )
                        {
                            $tabs[] = Tabs\Tab::make($locale)
                                ->schema([
                                    Forms\Components\TextInput::make('title_' . $locale),
                                    Forms\Components\TextArea::make('description_' . $locale),
                                ]);
                        }

                        return $tabs;
                    }),
```

Then on actions of table function:

```
->actions([
                Tables\Actions\EditAction::make()
                    ->mutateRecordDataUsing(function (Model $record, array $data): array {
                        
                        foreach($record->getTranslations() as $tfield => $translation)
                        {
                            foreach( $mylocales as $locale )
                            {
                                $data[$tfield . "_$locale"] = $translation[$locale] ?? "";
                            }
                        }

                        return $data;
                    })
                    ->using(function (Model $record, array $data): Model {
                        
                        foreach($record->getTranslations() as $tfield => $translation)
                        {
                            foreach( $mylocales as $locale )
                            {
                                $data[$tfield][$locale] = $data[$tfield . "_$locale"] ?? "";
                                unset($data[$tfield . "_$locale"]);
                            }
                        }

                        $record->update($data);
                        return $record;
                    }),
```

**Reply by machacekmartin** [reply]
This works but is not as clean as it could be using traits.. sadly it currently seems to be the only option that works for me. When trying @tombien solution - I get same result as @ArtMin96
